File: lib/features/movies/domain/repositories/movie_repository.dart
import 'package:dartz/dartz.dart';
import 'package:movieswipe/core/errors/failures.dart';
import 'package:movieswipe/features/movies/domain/entities/movie.dart';

/// Abstract repository interface for movies
abstract class MovieRepository {
  /// Get list of movies
  Future<Either<Failure, List<Movie>>> getMovies();

  /// Get personalized movie recommendations
  Future<Either<Failure, List<Movie>>> getRecommendedMovies();

  /// Swipe movie (like/dislike)
  Future<Either<Failure, void>> swipeMovie(int movieId, bool isLike, String userId, {int? rating});

  /// Search movies
  Future<Either<Failure, List<Movie>>> searchMovies(String query);
}

File: backend/app/domain/repositories/movie_repository.py
"""Movie repository interface (abstract)."""

from abc import ABC, abstractmethod
from typing import List
from app.domain.entities.movie import Movie


class MovieRepository(ABC):
    """Abstract repository for Movie entity."""

    @abstractmethod
    def get_all(self) -> List[Movie]:
        """Get all movies."""
        pass

    @abstractmethod
    def get_by_id(self, movie_id: int) -> Movie | None:
        """Get movie by ID."""
        pass

    @abstractmethod
    def create(self, movie: Movie) -> Movie:
        """Create a new movie."""
        pass

    @abstractmethod
    def swipe(self, movie_id: int, is_like: bool, user_id: str, rating: int | None = None) -> None:
        """Record a swipe action."""
        pass

    @abstractmethod
    def delete_swipe(self, movie_id: int, user_id: str) -> None:
        """Delete a swipe record (unlike/unpass)."""
        pass
